doc/design/memory.md

```diff
@@ -0,0 +1,1595 @@
+  last value is allocated in the space.  The resulting exposure encompasses all
+  allocation times.
+- A non-`NULL` `to` pointer to the associated tospace indicates the space is a
+  fromspace; a non-`NULL` `fr` pointer to the associated fromspace indicates the
+  space is a tospace.
+- A non-`NULL` `lazy` pointer indicates one or more lazy suspensions stored in
+  the space were forced, and the resulting values are stored in the lazy spaces.
+  These values are considered to be the same ages as the suspensions which
+  generated them, and they are post-order recursively copied into place with
+  their creators during copy-collection in order to preserve the value ordering
+  invariant.
+- The `frontier` pointer demarcates the boundary between values and wilderness.
+  Total bytes allocated is the difference between the metadata base address and
+  `frontier`.
+- The `unreachable` stack of deallocated values is used for incremental
+  decrefing.  Threading a per space stack through unreachable deallocated values
+  allows us to deallocate one value at a time and incrementally perform
+  cascading decrefs.
+- The `allocated` number of bytes is increased to incorporate newly allocated
+  values at the same time as `frontier` is bumped, and it is decreased as values
+  are deallocated as a side effect of decrefing.  If `allocated` drops to 0 and
+  the space has no associated fromspace, it can be destroyed as soon as its
+  unreachable stack is empty.  In practice this seldom occurs because
+  copy-collection commences well before the space becomes empty.
+- The space is capable of containing `size` bytes of value data.  Space
+  utilization is `allocated / size`; copy-collection commences when utilization
+  drops below a chosen threshold.
+
+### Global value layout
+
+Global and local values have the same layout, except that global value headers
+are prefixed with an additional word for refcounting metadata.  Following are
+the valid states for the global value header word.
+
+Global value header bit encodings:
+
+- **c:** Reference count.
+- **z:** Finalizable if 1.  The finalizer function is a low-level runtime
+  function associated with the type, and resurrection is impossible.  Immutable
+  value finalization is intended primarily for runtime implementation of file
+  descriptor management and other similar use cases, rather than as a Hemlock
+  language facility.
+- **s:** Non-`NULL` per space `unreachable` stack linkage.
+- **f:** Forwarding pointer, overlaid with reference count so that the
+  transition from refcounting to forwarding is atomic.
+
+**Global value header word:**
+
+```
+Non-zero refcount:
+
+  cccccccc cccccccc cccccccc cccccccc cccccccc cccccccc cccccccc ccccc0z1
+
+Forwarding pointer:
+
+  ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff fffff000
+
+Zero refcount, deallocated, in space's unreachable value stack:
+
+  ssssssss ssssssss ssssssss ssssssss ssssssss ssssssss ssssssss sssss0z0
```

Comment:
Just to make sure I understand this. Whenever a value is decrefed, to zero, we take over the `s` value portion of this header and make it point to the old top of the unreachable value stack. At the same time, we make the space metadata header point to this current header as the top of the unreachable value stack? At some point, the GC will come along and evaluate the actual value contents associated with this header to figure out if it points to any other values it needs to decref. Once the GC is done with that, it zeros out this entire header.

Reply:
Yes, that is correct.  Although we could immediately decref and deallocate a limited number of values (and we probably *will* perform one eager round of decref/deallocate), a long reference chain can trigger an arbitrarily long sequence of decref/deallocate operations.  This stack allows us to make such work incremental.